rlclientlib/dedup.cc

```diff
@@ -1,83 +1,315 @@
+}
+
+float dedup_state::get_ewma_value() const {
+  return _ewma.value();
+}
+
+int dedup_state::get_all_values(const std::unordered_map<generic_event::object_id_t, size_t>& src, generic_event::object_list_t& action_ids, std::vector<string_view>& action_values, api_status* status)
+{
+  std::unique_lock<std::mutex> mlock(_mutex);
+
+  for(auto a : src) {
+    auto content = _dict.get_object(a.first);
+    if(content.size() == 0) {
+      RETURN_ERROR_LS(nullptr, status, compression_error) << "Key not found while building batch dictionary";
+    }
+    action_ids.push_back(a.first);
+    action_values.push_back(content);
+  }
+
+  return error_code::success;
+}
+
+int dedup_state::remove_all_values(const std::unordered_map<generic_event::object_id_t, size_t>&input, api_status* status) {
+  std::unique_lock<std::mutex> mlock(_mutex);
+  for(auto kv: input) {
+    if(!_dict.remove_object(kv.first, kv.second)) {
+      RETURN_ERROR_LS(nullptr, status, compression_error) << "Key not found while prunning dedup_dict";
+    }
+  }
+
+  return error_code::success;
+}
+
+void dedup_state::update_ewma(float value)
+{
+  //This is racy, but update only reads and modifies a single value, so it won't lead to data corruption 
+  _ewma.update(value);
+}
+
+int dedup_state::compress(generic_event::payload_buffer_t& input, api_status* status) const {
+  //XXX this method is only const because we're not reusing the compression context
+  return _compressor.compress(input, status);
+}
+
+int dedup_state::transform_payload_and_add_objects(const char* payload, std::string& edited_payload, generic_event::object_list_t& object_ids, api_status* status){
+  std::unique_lock<std::mutex> mlock(_mutex);
+  return _dict.transform_payload_and_add_objects(payload, edited_payload, object_ids, status);
+}
+
+action_dict_builder::action_dict_builder(dedup_state& state):
+  _size_estimate(0)
+  , _state(state) {}
+
+void action_dict_builder::add(const generic_event::object_list_t &object_ids)
+{
+  for(auto aid : object_ids) {
+    auto it = _used_objects.find(aid);
+    if (it == _used_objects.end())
     {
-      _entries.insert({ hash, dict_entry(start, length) });
+      //XXX FAIL had if the action is missing
+      auto content = _state.get_object(aid);
+      _used_objects.insert({ aid, 1 });
+      _size_estimate += sizeof(size_t) + content.size();
     }
     else
     {
-      ++it->second._count;
+      ++it->second;
     }
-    return hash;
   }
+}
 
-  bool dedup_dict::remove_object(generic_event::object_id_t oid)
-  {
-    auto it = _entries.find(oid);
-    if (it == _entries.end())
-      return false;
+size_t action_dict_builder::size() const
+{
+  return _size_estimate * _state.get_ewma_value();
+}
+
+int action_dict_builder::finalize(generic_event& evt, api_status* status)
+{
+  i_time_provider* _time_provider = nullptr;
+  l::dedup_info_serializer ser;
+  const auto now = _time_provider != nullptr ? _time_provider->gmt_now() : timestamp();
```

Comment:
how can `_time_provider` not be `nullptr` if it is just set to `nullptr` two lines above?

Reply:
This is a TODO that slipped over, good catch.